safehttp/plugins/xsrf/xsrf.go

```diff
@@ -34,40 +35,64 @@ type UserIdentifier interface {
 }
 
 // Interceptor implements XSRF protection. It requires an application key and a
-// storage service. The appKey uniquely identifies each registered service and
+// storage service. The AppKey uniquely identifies each registered service and
 // should have high entropy. The storage service supports retrieving ID's of the
-// application's users. Both the appKey and user ID are used in the XSRF
-// token generation algorithm.
+// application's users. Both the AppKey and user ID are used in the XSRF
+// token generation and verification algorithm.
 type Interceptor struct {
 	AppKey     string
 	Identifier UserIdentifier
 }
 
+type tokenCtxKey struct{}
+
+// Token returns the XSRF token from the safehttp.IncomingRequest Context, if
+// present, and, otherwise, returns an empty string.
+func Token(r *safehttp.IncomingRequest) string {
+	tok := r.Context().Value(tokenCtxKey{})
+	if tok == nil {
+		return ""
+	}
+	return tok.(string)
+}
+
 // Before should be executed before directing the safehttp.IncomingRequest to
-// the handler to ensure it is not part of the Cross Site Request
-// Forgery. It checks for the presence of an xsrf-token in the request body and
-// validates it based on the userID associated with the request.
-func (p *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
-	userID, err := p.Identifier.UserID(r)
+// the handler to ensure it is not part of the Cross-Site Request
+// Forgery. In case of state changing methods, it checks for the
+// presence of an xsrf-token in the request body and validates it based on the
+// userID associated with the request. It also adds a cryptographically safe
+// XSRF token to the safehttp.IncomingRequest context.
+func (i *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	userID, err := i.Identifier.UserID(r)
 	if err != nil {
 		return w.ClientError(safehttp.StatusUnauthorized)
 	}
-	f, err := r.PostForm()
-	if err != nil {
-		mf, err := r.MultipartForm(32 << 20)
+
+	tok := xsrftoken.Generate(i.AppKey, userID, r.URL.String())
+	r.SetContext(context.WithValue(r.Context(), tokenCtxKey{}, tok))
+
+	stateChangingMethods := map[string]bool{
+		safehttp.MethodPost:  true,
+		safehttp.MethodPatch: true,
+	}
+	if m := r.Method(); stateChangingMethods[m] {
```

Comment:
AFAIU, you can't make PUT and DELETE requests  from HTML forms. We also don't allow cross-site state changing requests unless directed to an endpoint where CORS is specifically allowed, in which case the Fetch Metedata policies don't take effect. So the only requests we expect to contain tokens in the body are POST and PATCH requests.

Reply:
Right, but can't you make PUT and DELETE requests using JS and `fetch` to your own site. Shouldn't those requests contain XSRF tokens? Maybe not, because if they would be made cross-site then that would trigger a CORS preflight(?), which we won't handle.

I don't feel like I have enough context to judge this, so I trust what you've made. :) :+1: If there is a problem then hopefully @empijei and @kele will notice it.